libtiledbsoma/src/utils/arrow_adapter.cc

```diff
@@ -1150,6 +1190,104 @@ ArraySchema ArrowAdapter::tiledb_schema_from_arrow_schema(
     return schema;
 }
 
+std::pair<Attribute, std::optional<Enumeration>>
+ArrowAdapter::tiledb_attribute_from_arrow_schema(
+    std::shared_ptr<Context> ctx,
+    ArrowSchema* arrow_schema,
+    std::string_view type_metadata,
+    PlatformConfig platform_config) {
+    auto type = ArrowAdapter::to_tiledb_format(arrow_schema->format);
+    if (strcmp(arrow_schema->name, "soma_geometry") == 0) {
+        if (type_metadata.compare("WKB") == 0) {
+            type = TILEDB_GEOM_WKB;
+        } else {
+            throw std::runtime_error(
+                std::string("Unkwown type metadata for `soma_geometry`. "
+                            "Expected 'WKB', found ") +
+                type_metadata.data());
+        }
+    }
+
+    Attribute attr(*ctx, arrow_schema->name, type);
+
+    FilterList filter_list = ArrowAdapter::_create_attr_filter_list(
+        arrow_schema->name, platform_config, ctx);
+    attr.set_filter_list(filter_list);
+
+    if (arrow_schema->flags & ARROW_FLAG_NULLABLE) {
+        attr.set_nullable(true);
+    }
+
+    if (ArrowAdapter::arrow_is_var_length_type(arrow_schema->format)) {
+        attr.set_cell_val_num(TILEDB_VAR_NUM);
+    }
+
+    std::optional<Enumeration> enmr = std::nullopt;
+
+    if (arrow_schema->dictionary != nullptr) {
+        auto enmr_format = arrow_schema->dictionary->format;
+        auto enmr_type = ArrowAdapter::to_tiledb_format(enmr_format);
+        enmr = Enumeration::create_empty(
+            *ctx,
+            arrow_schema->name,
+            enmr_type,
+            ArrowAdapter::arrow_is_var_length_type(enmr_format) ?
+                TILEDB_VAR_NUM :
+                1,
+            arrow_schema->flags & ARROW_FLAG_DICTIONARY_ORDERED);
+        AttributeExperimental::set_enumeration_name(
+            *ctx, attr, arrow_schema->name);
+        LOG_DEBUG(fmt::format(
+            "[ArrowAdapter] dictionary for {} as {} {}",
+            std::string(arrow_schema->name),
+            tiledb::impl::type_to_str(enmr_type),
+            std::string(enmr_format)));
+    }
+
+    return {attr, enmr};
+}
+
+std::pair<Dimension, bool> ArrowAdapter::tiledb_dimension_from_arrow_schema(
```

Comment:
i'd prefer you move the body of this dimension function above the body of the corresponding attribute funtion